### PR TITLE
[nbc] Add support for series playlists

### LINKED ISSUE #11104: Some NBC playlists not supported
## Please follow the guide below

- You will be asked some questions and requested to provide some information, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *issue* (like that [x])
- Use *Preview* tab to see how your issue will actually look like

---

### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2016.11.02*. If it's not read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [ x] I've **verified** and **I assure** that I'm running youtube-dl **2016.11.02**

### Before submitting an *issue* make sure you have:
- [x ] At least skimmed through [README](https://github.com/rg3/youtube-dl/blob/master/README.md) and **most notably** [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [x ] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones

### What is the purpose of your *issue*?
- [ ] Bug report (encountered problems with youtube-dl)
- [ x] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

Some playlists on NBC are not working
```
keybounceMBP:timeless michael$ youtube-dl -v http://www.nbc.com/timeless/episodes
[debug] System config: []
[debug] User config: [u'-o', u'%(title)s.%(ext)s', u'-f', u'\nbest[ext=mp4][height>431][height<=480]/\nbest[ext=mp4][height>431][height<=576]/\nbestvideo[ext=mp4][height=480]+bestaudio[ext=m4a]/\nbest[ext=mp4][height>340][height<=431]/\nbestvideo[ext=mp4][height>360][height<=576]+bestaudio/\nbest[height>340][height<=576]/\nbestvideo[height>360][height<=576]+bestaudio/\nbestvideo[height=360]+bestaudio/\nbest[ext=mp4][height>=280][height<=360]/\nbest[height<=576]/\nworst', u'--write-sub', u'--write-auto-sub', u'--sub-lang', u'en,enUS', u'--sub-format', u'srt/ass/best', u'--embed-subs', u'--recode-video', u'mp4', u'--mark-watched']
[debug] Command-line args: [u'-v', u'http://www.nbc.com/timeless/episodes']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2016.11.02
[debug] Python version 2.7.5 - Darwin-13.4.0-x86_64-i386-64bit
[debug] exe versions: ffmpeg 3.2, ffprobe 3.2, rtmpdump 2.4
[debug] Proxy map: {}
[generic] episodes: Requesting header
WARNING: Falling back on generic information extractor.
[generic] episodes: Downloading webpage
[generic] episodes: Extracting information
ERROR: Unsupported URL: http://www.nbc.com/timeless/episodes
Traceback (most recent call last):
  File "/Users/michael/bin/youtube-dl/youtube_dl/extractor/generic.py", line 1634, in _real_extract
    doc = compat_etree_fromstring(webpage.encode('utf-8'))
  File "/Users/michael/bin/youtube-dl/youtube_dl/compat.py", line 2525, in compat_etree_fromstring
    doc = _XML(text, parser=etree.XMLParser(target=_TreeBuilder(element_factory=_element_factory)))
  File "/Users/michael/bin/youtube-dl/youtube_dl/compat.py", line 2514, in _XML
    parser.feed(text)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/xml/etree/ElementTree.py", line 1642, in feed
    self._raiseerror(v)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/xml/etree/ElementTree.py", line 1506, in _raiseerror
    raise err
ParseError: not well-formed (invalid token): line 3, column 47
Traceback (most recent call last):
  File "/Users/michael/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 694, in extract_info
    ie_result = ie.extract(url)
  File "/Users/michael/bin/youtube-dl/youtube_dl/extractor/common.py", line 357, in extract
    return self._real_extract(url)
  File "/Users/michael/bin/youtube-dl/youtube_dl/extractor/generic.py", line 2417, in _real_extract
    raise UnsupportedError(url)
UnsupportedError: Unsupported URL: http://www.nbc.com/timeless/episodes

keybounceMBP:timeless michael$ 
```

The individual episodes on that page do download, one at a time.

### PR DESCRIPTION
**Pull request by ppritcha**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

The current NBC extractor doesn't support creating playlists for the episodes of TV series.  This pull request adds the ability to extract a playlist of every season and episode from an episodes page.  Example page: https://www.nbc.com/saved-by-the-bell/episodes

Closes #11104.

### CODE REVIEW

**Comment by ppritcha** [comment]
Hi youtube-dl maintainers!  Is there anything further that you need from me for this this pull request?  It's been open a couple of weeks now, and passes the Travis checks.

The playlist capability will allow us to automatically download new episodes of TV shows as they are released, using a cron or scheduler script.  They are typically released on the web the day after they are broadcast.  This network is one of the "big three" TV networks in the U.S, with millions of viewers, so it's a very useful extractor for youtube-dl.

**Comment by dstftw** [comment]
Read coding conventions and fix code.

**Comment by ppritcha** [comment]
Thanks for the feedback.  I had previously read the coding conventions, and thought I was following them, so I don't know where the code is out of compliance with them.  Can you give me a hint?

File: youtube_dl/extractor/extractors.py [hunk 1]
```diff
@@ -1367,7 +1368,7 @@
     WebOfStoriesPlaylistIE,
 )
 from .weibo import (
-    WeiboIE, 
+    WeiboIE,
```
Comment (dstftw) [inline youtube_dl/extractor/extractors.py]: Don't mix unrelated changes in single PR.

File: youtube_dl/extractor/nbc.py [hunk 2]
```diff
@@ -15,7 +15,103 @@
...
+    urljoin,
+    parse_duration,
 )
+import datetime
+
+
+class NBCPlaylistIE(InfoExtractor):
+    IE_DESC = 'NBC series playlists'
+    IE_NAME = 'nbc.com:playlist'
+    _VALID_URL = r'(?i)https?://(?:www\.)nbc.com/(?P<id>[\w-]+)/episodes?$'
```
Comment (dstftw) [inline youtube_dl/extractor/nbc.py]: No `(?i)`, no `$`. Dots must be escaped.

File: youtube_dl/extractor/nbc.py [hunk 3]
```diff
@@ -15,7 +15,103 @@
...
+        },
+    ]
+
+    def extract_episode_info(self, url, json_data):
+
+        series_title = ''
+        included = json_data['included']
+        for inc in included:
+            if inc.get('type') == 'shows':
+                series_title = inc.get('attributes', {}).get('shortTitle')
```
Comment (dstftw) [inline youtube_dl/extractor/nbc.py]: Breaks.

File: youtube_dl/extractor/nbc.py [hunk 4]
```diff
@@ -15,7 +15,103 @@
...
+
+        entries = []
+        for ep in json_data['data']:
+            type = ep.get('type')
+            if type == 'videos':
+                atr = ep.get('attributes')
+                if atr:
+                    image_url = ''
+                    image_id = ep.get('relationships', {}).get('image', {}).get('data', {}).get('id')
```
Comment (dstftw) [inline youtube_dl/extractor/nbc.py]: Breaks.

File: youtube_dl/extractor/nbc.py [hunk 5]
```diff
@@ -15,7 +15,103 @@
...
+                    if episode_url:
+                        entries.append({
+                            '_type': 'url',
+                            'id': ep.get('id'),
+                            'ie_key': 'NBC',
+                            'title': series_title,
+                            'url': episode_url,
+                            'duration': parse_duration(atr.get('runTime')),
+                            'thumbnail': image_url,
+                            'upload_date': datetime.datetime.strptime(atr.get('airdate')[:19], '%Y-%m-%dT%H:%M:%S').strftime('%Y%m%d'),
```
Comment (dstftw) [inline youtube_dl/extractor/nbc.py]: Breaks.

File: youtube_dl/extractor/nbc.py [hunk 6]
```diff
@@ -15,7 +15,103 @@
...
+            'url': 'https://www.nbc.com/heroes/episodes',
+            'info_dict': {
+                'id': '99d3a2c1-fd98-43b9-a7a4-f7872b0eb808',
+                'title': 'Heroes',
+            },
+            'playlist_count': 78,
+        },
+    ]
+
+    def extract_episode_info(self, url, json_data):
```
Comment (dstftw) [inline youtube_dl/extractor/nbc.py]: Inline.

File: youtube_dl/extractor/nbc.py [hunk 7]
```diff
@@ -15,7 +15,103 @@
     unescapeHTML,
     update_url_query,
     int_or_none,
+    url_or_none,
+    js_to_json,
+    urljoin,
+    parse_duration,
```
Comment (dstftw) [inline youtube_dl/extractor/nbc.py]: Alphabetic.

File: youtube_dl/extractor/nbc.py [hunk 8]
```diff
@@ -15,7 +15,103 @@
...
+                        entries.append({
+                            '_type': 'url',
+                            'id': ep.get('id'),
+                            'ie_key': 'NBC',
+                            'title': series_title,
+                            'url': episode_url,
+                            'duration': parse_duration(atr.get('runTime')),
+                            'thumbnail': image_url,
+                            'upload_date': datetime.datetime.strptime(atr.get('airdate')[:19], '%Y-%m-%dT%H:%M:%S').strftime('%Y%m%d'),
+                            'episode_title': atr.get('title'),
```
Comment (dstftw) [inline youtube_dl/extractor/nbc.py]: No such meta field.

File: youtube_dl/extractor/nbc.py [hunk 9]
```diff
@@ -15,7 +15,103 @@
...
+
+        show_name_js = self._search_regex(
+            r'<script>PRELOAD=(.*)</script>', webpage, 'show_name')
+        show = self._parse_json(show_name_js, show_name,
+                                transform_source=js_to_json)
+        show_id = show.get('xref', {}).get('/' + show_name)
+
+        entries = []
+        if show_id:
+            episodes_url = urljoin(
```
Comment (dstftw) [inline youtube_dl/extractor/nbc.py]: What are you even trying to do here?

File: youtube_dl/extractor/nbc.py [hunk 10]
```diff
@@ -15,7 +15,103 @@
...
+                url, 'https://api.nbc.com/v3.14/videos?include=image%%2Cshow.image%%2Cshow.aggregates&filter%%5Bshow%%5D=%s&filter%%5Btype%%5D%%5Bvalue%%5D=Full%%20Episode&filter%%5Btype%%5D%%5Boperator%%5D=%%3D&page%%5Bnumber%%5D=1&page%%5Bsize%%5D=50&sort=-airdate' % show_id)
```
Comment (dstftw) [inline youtube_dl/extractor/nbc.py]: `update_url_query`.

File: youtube_dl/extractor/nbc.py [hunk 11]
```diff
@@ -15,7 +15,103 @@
...
+            while (episodes_url):
+                json_data = self._download_json(
+                    episodes_url, 'Downloading episode playlist')
+                result = self.extract_episode_info(url, json_data)
+                entries += result['episodes']
+                series_title = result['series_title']
+                episodes_url = json_data.get('links', {}).get('next')
```
Comment (dstftw) [inline youtube_dl/extractor/nbc.py]: Breaks.

File: youtube_dl/extractor/nbc.py [hunk 12]
```diff
@@ -15,7 +15,103 @@
...
+                        })
+
+        return {'series_title': series_title, 'episodes': entries}
+
+    def _real_extract(self, url):
+        show_name = self._match_id(url)
+        webpage = self._download_webpage(url, show_name)
+
+        show_name_js = self._search_regex(
+            r'<script>PRELOAD=(.*)</script>', webpage, 'show_name')
```
Comment (dstftw) [inline youtube_dl/extractor/nbc.py]: If you expect dict then capture a dict.

File: youtube_dl/extractor/nbc.py [hunk 13]
```diff
@@ -214,7 +310,8 @@ def _real_extract(self, url):
                 break
         else:
             source_url = video_source['ottStreamUrl']
-        is_live = video_source.get('type') == 'live' or video_source.get('status') == 'Live'
+        is_live = video_source.get(
+            'type') == 'live' or video_source.get('status') == 'Live'
```
Comment (dstftw) [inline youtube_dl/extractor/nbc.py]: Don't touch unrelated code.

File: youtube_dl/extractor/nbc.py [hunk 14]
```diff
@@ -377,7 +474,8 @@ def _real_extract(self, url):
         mobj = re.match(self._VALID_URL, url)
         video_id = mobj.group('id')
         if video_id is not None:
-            all_info = self._download_xml('http://www.nbcnews.com/id/%s/displaymode/1219' % video_id, video_id)
+            all_info = self._download_xml(
+                'http://www.nbcnews.com/id/%s/displaymode/1219' % video_id, video_id)
```
Comment (dstftw) [inline youtube_dl/extractor/nbc.py]: Don't touch unrelated code.

File: youtube_dl/extractor/nbc.py [hunk 15]
```diff
@@ -15,7 +15,103 @@
...
+                                inc_atr = inc.get('attributes')
+                                if inc_atr:
+                                    image_url = url_or_none(urljoin(url, inc_atr.get('path')))
+                    episode_url = url_or_none(atr.get('fullUrl'))
+                    if episode_url:
+                        entries.append({
+                            '_type': 'url',
+                            'id': ep.get('id'),
+                            'ie_key': 'NBC',
+                            'title': series_title,
```
Comment (dstftw) [inline youtube_dl/extractor/nbc.py]: For `url` type any metadata here have no effect.

**Comment by ppritcha** [comment]
Thanks very much for the detailed feedback.  I'll try to figure out what you mean in each case and make appropriate changes.  On a side note, I don't think these are covered in your coding conventions documentation, so that may be an area that needs some more detail or examples.  I'd be willing to help make changes to that if you will accept that help.
